### PR TITLE
Fix missing submodule update after rebase, allow more customized .gitsubprojects

### PR DESCRIPTION
**Pull request by tribal-tec**

### CODE REVIEW

File: GitExternal.cmake [hunk 1]
```diff
@@ -112,6 +112,9 @@ function(GIT_EXTERNAL DIR REPO tag)
       RESULT_VARIABLE nok ERROR_VARIABLE error WORKING_DIRECTORY "${DIR}")
     if(nok)
       message(FATAL_ERROR "git checkout ${TAG} in ${DIR} failed: ${error}\n")
+    else()
+      execute_process(COMMAND "${GIT_EXECUTABLE}" submodule update --init
```
Comment (eile) [inline GitExternal.cmake]: add --recursive

File: GitExternal.cmake [hunk 2]
```diff
@@ -180,8 +183,7 @@ function(GIT_EXTERNAL DIR REPO tag)
       "  RESULT_VARIABLE nok ERROR_VARIABLE error\n"
       "  WORKING_DIRECTORY \"${DIR}\")\n"
       "if(nok)\n"
-      "  message(FATAL_ERROR \"git checkout ${TAG} in ${__dir} failed: ${error}\n\")\n"
-      "endif()\n"
+      "  message(FATAL_ERROR \"git checkout ${TAG} in ${__dir} failed: \${error}\")\n"
     )
```
Comment (tribal-tec) [inline GitExternal.cmake]: No, continues with else()endif() in l.208

File: GitExternal.cmake (hunk 1 above)
Comment (tribal-tec) [inline GitExternal.cmake]: Done

File: GitExternal.cmake (hunk 2 above)
Comment (eile) [inline GitExternal.cmake]: Yes, reviewing offline now. Much easier to see.

File: SubProject.cmake [hunk 3]
```diff
@@ -51,12 +51,10 @@
...
 include(${CMAKE_CURRENT_LIST_DIR}/CommonGraph.cmake)
 
-if(TARGET git_subproject_${PROJECT_NAME}_done)
+if(GIT_SUBPROJECT_DONE_${PROJECT_NAME})
   return()
 endif()
-add_custom_target(git_subproject_${PROJECT_NAME}_done)
-set_target_properties(git_subproject_${PROJECT_NAME}_done PROPERTIES
-  EXCLUDE_FROM_DEFAULT_BUILD ON FOLDER ${PROJECT_NAME}/zzphony)
+set(GIT_SUBPROJECT_DONE_${PROJECT_NAME} ON)
```
Comment (eile) [inline SubProject.cmake]: That does not have global scope as the target does.

Comment (tribal-tec) [inline SubProject.cmake]: target is still the wrong thing here. let's see offline

File: GitExternal.cmake [hunk 4]
```diff
@@ -112,6 +112,9 @@ function(GIT_EXTERNAL DIR REPO tag)
       RESULT_VARIABLE nok ERROR_VARIABLE error WORKING_DIRECTORY "${DIR}")
     if(nok)
       message(FATAL_ERROR "git checkout ${TAG} in ${DIR} failed: ${error}\n")
+    else()
+      execute_process(COMMAND "${GIT_EXECUTABLE}" submodule update --init --recursive
+        WORKING_DIRECTORY "${DIR}")
```
Comment (hernando) [inline GitExternal.cmake]: Do you think this solves the arbitrary failures with parallel clones?

File: GitExternal.cmake [hunk 5]
```diff
@@ -112,6 +112,9 @@ function(GIT_EXTERNAL DIR REPO tag)
       RESULT_VARIABLE nok ERROR_VARIABLE error WORKING_DIRECTORY "${DIR}")
     if(nok)
       message(FATAL_ERROR "git checkout ${TAG} in ${DIR} failed: ${error}\n")
```
Comment (hernando) [inline GitExternal.cmake]: return after fatal errors?
Why in line 103 the message is not a fatal error?

File: SubProject.cmake [hunk 6]
```diff
@@ -157,11 +155,12 @@ if(EXISTS "${CMAKE_CURRENT_SOURCE_DIR}/.gitsubprojects")
     set(GIT_SUBPROJECTS_SCRIPT
       "${CMAKE_CURRENT_BINARY_DIR}/UpdateSubprojects.cmake")
     file(WRITE "${GIT_SUBPROJECTS_SCRIPT}"
-      "file(WRITE .gitsubprojects \"# -*- mode: cmake -*-\n\")\n")
+      "file(READ .gitsubprojects GIT_SUBPROJECTS_FILE)\n")
```
Comment (hernando) [inline SubProject.cmake]: What if the file doesn't exist at all (e.g. a project without subprojects)?

Comment (tribal-tec) [inline SubProject.cmake]: l.151: if(EXISTS "${CMAKE_CURRENT_SOURCE_DIR}/.gitsubprojects")

File: GitExternal.cmake (hunk 4 above)
Comment (tribal-tec) [inline GitExternal.cmake]: Maybe. You can retry if you want :)

File: GitExternal.cmake (hunk 5 above)
Comment (tribal-tec) [inline GitExternal.cmake]: in 103 it's not an error for optional externals. in 106 there's a fatal error. In doubt ask the original author :)

File: SubProject.cmake (hunk 3 above)
Comment (tribal-tec) [inline SubProject.cmake]: Done

File: SubProject.cmake (hunk 6 above)
Comment (hernando) [inline SubProject.cmake]: That was not visible in the diff :)

File: GitExternal.cmake (hunk 4 above)
Comment (hernando) [inline GitExternal.cmake]: If I don't have anything more important to do I'll try :P

**Pull request merged**